ipapython/fqdn.py

```diff
@@ -0,0 +1,29 @@
+#
+# Copyright (C) 2020  FreeIPA Contributors see COPYING for license
+#
+"""Get host's FQDN
+"""
+import socket
+
+
+def gethostfqdn():
+    """Get the fully qualified domain name of current host from glibc
+
+    :return: FQDN as str
+    """
+    hostname = socket.gethostname()
+    # this call can never fail except for misconfigured nsswitch.conf
+    # without nss-myhostname provider. The myhostname provider translates
+    # gethostname() to local interfaces.
+    gai = socket.getaddrinfo(
+        hostname,
+        None,  # service/port is irrelevant
+        family=socket.AF_UNSPEC,  # IPv4 or IPv6
+        type=socket.SOCK_DGRAM,  # optimization, TCP/RAW gives same result
+        # include canonical name in first addrinfo struct
+        # only use address family when at least one non-local interface
+        # is configured with that address family
+        flags=socket.AI_CANONNAME | socket.AI_ADDRCONFIG
+    )
+    # first addrinfo struct, fourth field is canonical name
+    return gai[0][3]
```

Comment:
We should check `len(gai) > 0` and either fail with a more descriptive exception than `IndexError`, or return `hostname` (even though it may not be fully qualified).  If nssswitch is misconfigured, we should at least try and be helpful instead of failing with a cryptic error :)

Reply:
I don't see how ``gai`` can be empty at this point. ``getaddrinfo()`` either returns at least one result or it fails with an error. All C code I have read recently always assumes that the result struct contains valid data on success.

Reply:
OK.  I prefer to be defensive ("if the return type admits it, I must handle it"), but I will leave it up to you.

Reply:
I have added a comment that explains why this line cannot fail.